src/mpi/romio/adio/common/ad_darray.c

```diff
@@ -106,29 +106,30 @@ int ADIO_Type_create_darray(int size, int rank, int ndims,
         }
 
         /* add displacement and UB */
-        disps[1] = st_offsets[ndims - 1];
+        disps[0] = st_offsets[ndims - 1];
         tmp_size = 1;
         for (i = ndims - 2; i >= 0; i--) {
             tmp_size *= array_of_gsizes[i + 1];
-            disps[1] += (MPI_Aint) tmp_size *st_offsets[i];
+            disps[0] += (MPI_Aint) tmp_size *st_offsets[i];
         }
     }
 
-    disps[1] *= orig_extent;
+    disps[0] *= orig_extent;
 
-    disps[2] = orig_extent;
+    lb = 0;
+    ub = orig_extent;
     for (i = 0; i < ndims; i++)
-        disps[2] *= (MPI_Aint) array_of_gsizes[i];
+        ub *= (MPI_Aint) array_of_gsizes[i];
 
-    disps[0] = 0;
-    blklens[0] = blklens[1] = blklens[2] = 1;
-    types[0] = MPI_LB;
-    types[1] = type_new;
-    types[2] = MPI_UB;
+    blklens[0] = 1;
+    types[0] = type_new;
 
```

Comment:
@roblatham00 Here is the usage that you are concerned with, right?

Reply:
not exactly:  the new code should be equivalent, except there's this strange notion of a "sticky" upper bound ((MPIR_Datatype_debug used to display the difference before MPI_LB/MPI_UB support was removed in 93475e42207 ).  Somewhere in the flattening code mess is logic that actually cares about sticky extent markers and this removal of MPI_UB  reminded me of the RESIZED vs STRUCT problem.

Reply:
Sorry I'm only now getting back to this PR.  I found some notes from my first version of a fix for HDF5: they had a type_contiguous containing a type_hindexed where the count used in the hindexed construction call was 0 (so the datatype size was 0).  So going into the type_contiguous combiner with 0-size components is where the "count" and "flatten" functions diverged.  And there was a semi-related case of type_contiguous with a count of 0 where count and flatten also diverged.

At first I spent a long time putting those two functions in sync with each other, and it always worked for whatever path I traced through for that bug, but later there would be another bug in another path.  So that's where I did the much simpler fix of simply not letting the "count" function have the final word, but letting "flatten" determine the final count itself.

I do however see a commit I never contributed back, that doesn't involve count-vs-flatten divergence, and does involve combiner_resize and LB... I'm going to look into that some more, and maybe it'll produce another contribution if I can figure out what testcase it was fixing.

Anyway my belief is the modernization changes shouldn't be having any effect on the flatten stuff, so I think this PR should still be good.

Reply:
@roblatham00 so I looked over my old PRs and did find one that was LB/UB related and separate from the flatten-vs-count divergence I talked about above:
    https://github.com/pmodels/mpich/pull/5302